### PR TITLE
[DML] Support convTranpose2d for DML

### PR DESCRIPTION
**Pull request by miaobin**
Rebase the code after PR #226 merged. Test test errors about convTranspose2d will all be resolved.
@fujunwei @mingmingtasd PTAL, thanks!

### CODE REVIEW

File: src/webnn_native/dml/GraphDML.cpp [hunk 1]
```diff
@@ -190,14 +190,43 @@ namespace webnn_native::dml {
             return {oStride, iStride, hStride, wStride};
         }
 
+        ::dml::TensorDimensions CalculateConvTranspose2dFilterLayoutStrides(
```
Comment (mingmingtasd) [inline src/webnn_native/dml/GraphDML.cpp:193]: How about renaming line 163 as CalculateConv2dFilterLayoutStrides() if it's only for conv2d?

Comment (miaobin) [reply src/webnn_native/dml/GraphDML.cpp:193]: Done.

File: src/webnn_native/dml/GraphDML.cpp [hunk 2]
```diff
@@ -234,6 +263,50 @@ namespace webnn_native::dml {
...
+            ::dml::TensorDimensions filterDims = filter.GetOutputDesc().sizes;
+            ::dml::TensorDimensions newFilterDims;
+            newFilterDims.resize(4);
+            switch (filterLayout) {
+                case wnn::ConvTranspose2dFilterOperandLayout::Iohw:
+                    newFilterDims[0] = filterDims[0];
+                    newFilterDims[1] = filterDims[1];
+                    newFilterDims[2] = filterDims[2];
+                    newFilterDims[3] = filterDims[3];
+                    filter = ::dml::Reinterpret(
```
Comment (fujunwei) [inline src/webnn_native/dml/GraphDML.cpp:278]: Combine L.278 / L.280 / L290 to one line in L307

File: src/webnn_native/dml/GraphDML.cpp [hunk 3]
```diff
@@ -412,6 +485,21 @@ namespace webnn_native::dml {
...
 
+        std::vector<const uint32_t> ImplicitPaddingforConvTranpose2d(
+            const ConvTranspose2dOptions* options,
+            ::dml::TensorDimensions inputDims,
+            ::dml::TensorDimensions filterDims) {
+            std::vector<int32_t> inputSize = {static_cast<int32_t>(inputDims[2]),
+                                              static_cast<int32_t>(inputDims[3])};
+            std::vector<int32_t> filterSize = {static_cast<int32_t>(filterDims[2]),
+                                               static_cast<int32_t>(filterDims[3])};
+            std::vector<int32_t> padding = utils::ComputeImplicitPaddingForConvTranspose2dAutoPad(
```
Comment (fujunwei) [inline src/webnn_native/dml/GraphDML.cpp:496]: can we change the function declaration like `std::vector<uint32_t> ComputeImplicitPaddingForConvTranspose2dAutoPad(const ConvTranspose2dOptions* options, std::vector<int32_t> inputDim, std::vector<int32_t> filterDim)`, so we can call it directly.

File: src/webnn_native/dml/GraphDML.cpp [hunk 4]
```diff
@@ -888,7 +976,82 @@ namespace webnn_native::dml {
...
+        std::vector<const uint32_t> endPaddingVector = {padding[1], padding[3]};
+        ::dml::Span<const uint32_t> endPadding(endPaddingVector);
+
+        ::dml::TensorDimensions outputShape = {};
+        outputShape.push_back(inputDims[0]);   // N
+        outputShape.push_back(filterDims[1]);  // C
+        if (options->outputSizes != nullptr) {
+            std::vector<int32_t> outputSizes(options->outputSizes,
+                                             options->outputSizes + options->outputSizesCount);
+            outputShape.push_back(outputSizes[0]);
```
Comment (fujunwei) [inline src/webnn_native/dml/GraphDML.cpp:1017]: You can define the uint32_t outputHeight and uint32_t outputWidth outside, so you can assign them like 
```
if (options->outputSizes != nullptr) {
  outputHeight  = options->outputSizes[0];
  outputWidth  = options->outputSizes[1];
} else {
  outputHeight  = 
  outputWidth  = 
}
outputShape.push_back(outputHeight);
outputShape.push_back(outputWidth);
```

File: src/webnn_native/dml/GraphDML.cpp [hunk 5]
```diff
@@ -888,7 +976,82 @@ namespace webnn_native::dml {
...
+            outputShape.push_back(outputWidth);
+        }
+
+        ::dml::Optional<::dml::Expression> bias = ::dml::NullOpt;
+        if (options->bias != nullptr) {
+            DAWN_ASSERT(mExpression.find(inputsOperand[2].Get()) != mExpression.end());
+            bias = mExpression.at(inputsOperand[2].Get());
+            ::dml::TensorDimensions biasDims = bias->GetOutputDesc().sizes;
+            if (biasDims[0] != filter.GetOutputDesc().sizes[0] || biasDims.size() != 1) {
+                return DAWN_INTERNAL_ERROR(
```
Comment (fujunwei) [inline src/webnn_native/dml/GraphDML.cpp:1036]: Use DAWN_INVALID_IF macro

**Review by fujunwei — APPROVED** [review]
LGTM with some comments.

File: src/webnn_native/dml/GraphDML.cpp (hunk 3 above)
Comment (miaobin) [reply src/webnn_native/dml/GraphDML.cpp:469]: This is a public method stored in `utils.cpp`, we will call it in both GraphIE and GraphDML, in GraphIE we use int32_t.

File: src/webnn_native/dml/GraphDML.cpp (hunk 4 above)
Comment (miaobin) [reply src/webnn_native/dml/GraphDML.cpp]: Thanks for the suggestion, I will pay more attention to some such problems that the code can be optimized in the future. PTAL, @fujunwei.

File: src/webnn_native/dml/GraphDML.cpp [hunk 6]
```diff
@@ -888,7 +947,79 @@ namespace webnn_native::dml {
...
+        } else {
+            outputHeight = strides[0] * (inputDims[2] - 1) + outputPadding[0] +
+                           ((filterDims[2] - 1) * dilations[0] + 1) - startPaddingVector[0] -
+                           endPaddingVector[0];
+            outputWidth = strides[1] * (inputDims[3] - 1) + outputPadding[1] +
+                          ((filterDims[3] - 1) * dilations[1] + 1) - startPaddingVector[1] -
+                          endPaddingVector[1];
+        }
+        outputShape.push_back(outputHeight);
+        outputShape.push_back(outputWidth);
```
Comment (mingmingtasd) [inline src/webnn_native/dml/GraphDML.cpp:998]: Could we use the calculated output shape from ConvTranspose2d.cpp?

Comment (miaobin) [reply src/webnn_native/dml/GraphDML.cpp]: Yes, I have optimized the code. PTAL, thanks!

File: src/webnn_native/dml/GraphDML.cpp (hunk 3 above)
Comment (fujunwei) [reply src/webnn_native/dml/GraphDML.cpp:469]: Can we make it as template for different data type?

Comment (miaobin) [reply src/webnn_native/dml/GraphDML.cpp]: Make the functions in Utils.h as template for different data type. Link to the [issue](Make the functions in Utils.h as template for different data type).
PTAL @fujunwei thanks!

File: src/webnn_native/dml/GraphDML.cpp [hunk 7]
```diff
@@ -412,6 +446,18 @@ namespace webnn_native::dml {
             return ImplicitPadding(options, input, {filterDims[2], filterDims[3]});
         }
 
+        std::vector<const uint32_t> ImplicitPaddingforConvTranpose2d(
+            const ConvTranspose2dOptions* options,
```
Comment (fujunwei) [inline src/webnn_native/dml/GraphDML.cpp:450]: Could we expand the function in the usage?

Comment (miaobin) [reply src/webnn_native/dml/GraphDML.cpp]: Done and rebased the code. PTAL, @fujunwei thanks a lot!

**Comment by fujunwei** [comment]
Thanks for interating patiently, i will merge it when the CI is green, and please roll the version to otcshare/Chromium, and make sure there are no regression for chromium prototype.

**Comment by fujunwei** [comment]
Merge it.

**Pull request merged**